### PR TITLE
Fix chrome/selenium tests

### LINKED ISSUE #25322: Test failures and hangs due to Selenium/Chrome issues
We have seen a lot of CI failures/hangs in due to issues with Chrome and Selenium. The investigation is ongoing but the affected tests have been disabled to unblock CI. This issue tracks fixing those tests and re-enabling them.

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Fixes #25322

### CODE REVIEW

**Review by pranavkm — APPROVED** [review]
#yolo

@javiercn

**Comment by dougbu** [comment]
🤞

**Comment by dougbu** [comment]
Just noticed the base branch. This is test-only and, if possible, we should have the Selenium tests running for 5.0

**Comment by JunTaoLuo** [comment]
I was planning on trying things out in master before back-porting the fixes to release/5.0.

File: eng/Versions.props [hunk 1]
```diff
@@ -266,10 +266,10 @@
     <NewtonsoftJsonBsonPackageVersion>1.0.2</NewtonsoftJsonBsonPackageVersion>
     <NewtonsoftJsonPackageVersion>12.0.2</NewtonsoftJsonPackageVersion>
     <NSwagApiDescriptionClientPackageVersion>13.0.4</NSwagApiDescriptionClientPackageVersion>
-    <SeleniumSupportPackageVersion>3.12.1</SeleniumSupportPackageVersion>
+    <SeleniumSupportPackageVersion>3.141.0</SeleniumSupportPackageVersion>
     <SeleniumWebDriverMicrosoftDriverPackageVersion>17.17134.0</SeleniumWebDriverMicrosoftDriverPackageVersion>
-    <SeleniumWebDriverChromeDriverPackageVersion>2.43.0</SeleniumWebDriverChromeDriverPackageVersion>
-    <SeleniumWebDriverPackageVersion>3.12.1</SeleniumWebDriverPackageVersion>
+    <SeleniumWebDriverChromeDriverPackageVersion>85.0.4183.8300</SeleniumWebDriverChromeDriverPackageVersion>
```
Comment (captainsafia) [inline eng/Versions.props:271]: This is a pretty big jump. Did something change with the way the package is versioning...

Comment (dougbu) [reply eng/Versions.props:271]: Yup, they moved to use versions matching Chrome versions. There as a jump from v.2.46.0 to v.71.0.3578.137 a while back.

In addition, we were almost 2 years out-of-date in using v2.43.0.

Comment (captainsafia) [reply eng/Versions.props:271]: I see. That makes sense.

How does this mesh with the fact that we don't control what version of Chrome is downloaded with the installer (AFAIK)?

Comment (dougbu) [reply eng/Versions.props:271]: It _should_ mean we can be proactive. Whenever a dev notices their Chrome installation wants to be restarted, they can check whether it's a new major version and update this particular package version. Or, @sebastienros could add a reminder to check the Selenium versions in his ops rotation invites every few (10?) rotations.

**Comment by dougbu** [comment]
> I was planning on trying things out in master before back-porting the fixes to release/5.0.

Makes sense.

Looks like you have at least one breaking change to handle. But, it's not that bad considering how far out of date we were.

File: eng/Versions.props (hunk 1 above)
Comment (sebastienros) [reply eng/Versions.props:271]: I could, I am afraid I will forget ;) I would suggest to update the operations book, and let's say "if your rotation happens on the first Tuesday of the month then you should ..."

Mondays are frequently skipped due to holidays

**Comment by Pilchie** [comment]
:eyes:

File: eng/Versions.props (hunk 1 above)
Comment (dougbu) [reply eng/Versions.props:271]: Good idea. May also need something in https://github.com/dotnet/aspnetcore-internal/blob/master/.github/ISSUE_TEMPLATE/Build_OPS_handoff.md @JunTaoLuo what do you think❔

**Comment by dougbu** [comment]
@JunTaoLuo I suggest copying #25323 into the 'release/3.1' branch if possible (@Pilchie and @wtgodbe is that branch open for test-only changes at this point❔). Eventually we'll want this fix there too because it also uses Selenium and our InstallGoogleChrome.ps1 script. E.g. some of the errors in my attempted validation of #25217 mention `cannot find Chrome binary`.

@captainsafia other validation problems w/ #25217 require your #25324 fix i.e. everyone will need to include your commit in their 3.1 PRs or validations are just busted. Sooner it can go in (too), the better. But, again, only if branch is open.

**Review by SteveSandersonMS — APPROVED** [review]
This looks great. Thanks for figuring this out, @JunTaoLuo!

**Comment by wtgodbe** [comment]
> (@wtgodbe is that branch open for test-only changes at this point❔)

3.1 isn't open for anything until patch tuesday in September when we whip 3.1.8

**Comment by JunTaoLuo** [comment]
Sounds like there's no rush to unblock 3.1 right now. I'll just continue with making fixes here and cherry-pick them to release/5.0 and 3.1 when it's done.

**Comment by JunTaoLuo** [comment]
I've updated the APIs but I'll work with @ajaybhargavb's to test it on Monday.

**Comment by dougbu** [comment]
@JunTaoLuo how far are we from having a commit we can `cherry-pick` into 3.1 PRs to see if they're ready to go❔

**Comment by BrennanConroy** [comment]
This PR is failing the exact same way as before, unable to launch chrome and hanging.

**Comment by ajaybhargavb** [comment]
> This PR is failing the exact same way as before, unable to launch chrome and hanging.

Looks like it is still trying to use the old chrome driver (2.43.xxx) instead of the new `85.0.xxx`. I wonder where it is picking that from. cc @JunTaoLuo 
```
 OpenQA.Selenium.WebDriverException : unknown error: cannot find Chrome binary\n  (Driver info: chromedriver=2.43.600210 (68dcf5eebde37173d4027fa8635e332711d2874a),platform=Windows NT 10.0.17763 x86_64)
```

**Comment by ajaybhargavb** [comment]
@JunTaoLuo's latest commit fixed the Chrome not being found issue but it is now throwing a null-ref,
```
System.NullReferenceException : Object reference not set to an instance of an object.
   at OpenQA.Selenium.Remote.HttpCommandExecutor.HttpRequestInfo..ctor(Uri serverUri, Command commandToExecute, CommandInfo commandInfo)
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteLogs.GetLog(String logKind)
   at OpenQA.Selenium.WebDriverExtensions.GetBrowserLogs(IWebDriver driver, LogLevel level) in /_/src/Shared/E2ETesting/WebDriverExtensions.cs:line 20
   at Microsoft.AspNetCore.E2ETesting.WaitAssert.WaitAssertCore[TResult](IWebDriver driver, Func`1 assertion, TimeSpan timeout) in /_/src/Shared/E2ETesting/WaitAssert.cs:line 120
   at Microsoft.AspNetCore.E2ETesting.WaitAssert.Exists(IWebDriver driver, By finder, TimeSpan timeout) in /_/src/Shared/E2ETesting/WaitAssert.cs:line 64
   at Microsoft.AspNetCore.E2ETesting.WaitAssert.Exists(IWebDriver driver, By finder) in /_/src/Shared/E2ETesting/WaitAssert.cs:line 51
   at Templates.Test.BlazorWasmTemplateTest.TestBasicNavigation(String appName, Boolean usesAuth, Boolean skipFetchData) in /_/src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:line 525
   at Templates.Test.BlazorWasmTemplateTest.BlazorWasmStandaloneTemplate_Works() in /_/src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:line 73
```

**Comment by JunTaoLuo** [comment]
Turns out that getting browser logs is broken: https://github.com/SeleniumHQ/selenium/issues/7342. Seems like the issue is fixed in 4.0.0-alpha3 so I'll try that version and see what happens.

**Comment by JunTaoLuo** [comment]
Okay this got even further but now I'm seeing actual test failures in blazor templates and component e2e tests.

**Comment by JunTaoLuo** [comment]
Hey @javiercn, you've been recommended to take a look at the test failures. Can you investigate and see what's causing the test failures.

File: src/Shared/E2ETesting/BrowserFixture.cs [hunk 2]
```diff
@@ -234,55 +237,75 @@ private async Task<(IWebDriver browser, ILogs log)> CreateSauceBrowserAsync(stri
...
+                    break;
+                case "internet explorer":
+                    options = new InternetExplorerOptions();
+                    break;
+                case "microsoftedge":
+                    options = new EdgeOptions();
+                    break;
+                default:
+                    throw new InvalidOperationException($"Browser name {sauce.BrowserName} not recognized");
+            }
```
Comment (javiercn) [inline src/Shared/E2ETesting/BrowserFixture.cs:258]: Why do we need this? We ended up not going this route @ajaybhargavb isn't that the case?

We only need chrome I believe

Comment (ajaybhargavb) [reply src/Shared/E2ETesting/BrowserFixture.cs:258]: We only need Chrome for local and CI tests. But we use other browsers when running in SauceLabs. We need to pass in the correct options to SauceLabs in that case. `DesiredCapabilities` used to be enough but that is now obsoleted in the newer version.

Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs:258]: @ajaybhargavb but do we actually run in SauceLabs? I thought we discarded that idea. That's why I'm asking.

Comment (ajaybhargavb) [reply src/Shared/E2ETesting/BrowserFixture.cs:258]: We still want to. What we discarded was the idea to do real device testing on SauceLabs.

**Comment by JunTaoLuo** [comment]
FYI, let's backport this fix to 5.0-rc2 only once we are done fixing it in master, turns out it's too late for RC1.

**Comment by dougbu** [comment]
> FYI, let's backport this fix to 5.0-rc2 only once we are done fixing it in master, turns out it's too late for RC1.

Agreed but we also want to fix Selenium tests in release/3.1. The same-site fixes in @BrennanConroy and @captainsafia's #25324 aren't enough. Would rather not back-port #25323 into a servicing branch due to the loss of code coverage.

**Comment by halter73** [comment]
@mkArtakMSFT Do we still need to assign an investigator for this?

**Comment by dougbu** [comment]
> Do we still need to assign an investigator for this?

We _urgently_ need an investigator because we're seeing the same issues in release/3.1 and that has a short timeline. @javiercn do you have time to help w/ the WASM template tests❔

**Comment by JunTaoLuo** [comment]
I'll take a look tonight and see if I can find anything or at least narrow down the issue we're seeing. Also I'm curious if https://github.com/dotnet/aspnetcore/pull/25689 might contain some of the fixes that'll get this working.

**Comment by dougbu** [comment]
> I'm curious if #25689 might contain some of the fixes that'll get this working.

@SteveSandersonMS❔❔

**Comment by JunTaoLuo** [comment]
Seems like those didn't affect the test failures seen in this PR. I'll try to pin down the cause a little more clearly tomorrow.

**Comment by SteveSandersonMS** [comment]
> I'm curious if #25689 might contain some of the fixes that'll get this working.

> @SteveSandersonMS❔❔

I hope so! The changes in #25689 are enough to make all the Components E2E tests pass for me locally. I don't know about the templates tests though - that's a separate thing.

**Comment by JunTaoLuo** [comment]
It didn't seem to fix the Components.E2ETests actually, I'm going to look into why today.

**Comment by JunTaoLuo** [comment]
Turns out the browser logs are not working. I might just disable the specific tests that checks logs to unblock the rest of the tests enabled by this PR while I investigate.

**Comment by JunTaoLuo** [comment]
@dougbu @mkArtakMSFT I suppose we want these fixes in rc2 as well?

**Comment by Pilchie** [comment]
> @dougbu Doug Bunting FTE @mkArtakMSFT Artak Mkrtchyan FTE I suppose we want these fixes in rc2 as well?

Yes, please!

**Comment by dougbu** [comment]
And, release/3.1 if possible ❕

**Pull request merged**